rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +86,172 @@ struct @(spec.base_type.type)_
 {
   using Type = @(spec.base_type.type)_<ContainerAllocator>;
 
-@# constructors (with and without allocator)
-@[for (alloc_type, alloc_name) in [['', ''], ['const ContainerAllocator & ', '_alloc']]]@
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
 @{
-used_alloc_name = False
-leading_colon = ':'
+class Member:
+    def __init__(self, name):
+        self.name = name
+        self.default_value = None
+        self.zero_value = None
+        self.zero_need_array_override = False
+        self.type = None
+
+class CommonMemberSet:
+    def __init__(self):
+        self.members = []
+
+    def add_member(self, memb):
+        if not self.members:
+            self.members.append(memb)
+            return True
+        new_def_noneness = memb.default_value is None
+        last_def_noneness = self.members[-1].default_value is None
+        new_zero_noneness = memb.zero_value is None
+        last_zero_noneness = self.members[-1].zero_value is None
+        if (new_def_noneness == last_def_noneness) and (new_zero_noneness == last_zero_noneness):
+            self.members.append(memb)
+            return True
+        return False
+
+def default_cpp_value_from_type(type):
+    if type == 'string':
+        return ''
+    elif type in ['float32', 'float64']:
+        return '0.0'
+    elif type == 'bool':
+        return 'false';
+    return '0'
```

Comment:
You might want to move this Python code into a module for better readability, linting, deduplication, etc since they don't seem to contain anything message specific.

Reply:
Good point.  Done in 91b5f94